samplekotlin/src/main/java/com/example/chattutorial/ChannelActivity.kt

```diff
@@ -44,12 +44,21 @@ class ChannelActivity : AppCompatActivity(R.layout.activity_channel) {
                 is MessageListViewModel.Mode.Normal -> messageInputViewModel.resetThread()
             }
         }
-        // step 4 - let the message input know when we are editing a message
+
+        // step 4 - handle navigate up state
+        messageListViewModel.state.observe(this) {
+            when (it) {
+                is MessageListViewModel.State.NavigateUp -> finish()
+                else -> Unit
+            }
+        }
+
+        // step 5 - let the message input know when we are editing a message
```

Comment:
is there no way to simplify this back button handling stuff? @JcMinarro thoughts?
it takes up too much of this part of the tutorial. we're trying to explain how to build chat, not how to handle a back button, it's a bit distracting

Reply:
Is there a reason mode isn't part of state?